dali/pipeline/operators/util/external_source.cu

```diff
@@ -12,21 +12,54 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include <utility>
+#include <memory>
+
 #include "dali/pipeline/operators/util/external_source.h"
 
 namespace dali {
 
+template <>
+struct ExternalSource<GPUBackend>::RecycleFunctor {
+  RecycleFunctor() = default;
+  RecycleFunctor(const RecycleFunctor &) {
+    assert(!"Should never happen");
+  }
+  RecycleFunctor(RecycleFunctor &&) = default;
+
+  RecycleFunctor(
+      ExternalSource<GPUBackend> *owner,
+      uptr_cuda_event_list_type event,
+      uptr_tl_list_type ptr)
+  : owner(owner), event(std::move(event)), ptr(std::move(ptr)) {}
+
+  ExternalSource<GPUBackend> *owner;
+  uptr_cuda_event_list_type event;
+  uptr_tl_list_type ptr;
+  void operator()() {
+    owner->RecycleBuffer(ptr, &event);
+  }
+};
+
 template<>
 void ExternalSource<GPUBackend>::RunImpl(DeviceWorkspace *ws) {
-  DALI_ENFORCE(data_in_tl_, "Cannot feed non-contiguous data to GPU op.");
-
-  auto &output = ws->Output<GPUBackend>(0);
-  output.Copy(tl_data_, (ws->has_stream() ? ws->stream() : 0));
+  uptr_tl_list_type data;
+  uptr_cuda_event_list_type cuda_event;
   {
-    std::lock_guard<std::mutex> busy_lock(busy_m_);
-    busy_ = false;
+    std::unique_lock<std::mutex> busy_lock(busy_m_);
+    cv_.wait(busy_lock, [&data = data_in_tl_]{return !data.empty();});
+    auto is_data_in_tl = data_in_tl_.front();
+    data_in_tl_.pop_front();
+    DALI_ENFORCE(is_data_in_tl, "Cannot feed non-contiguous data to GPU op.");
+    data = tl_data_.PopFront();
+    cuda_event = std::move(cuda_events_.GetEmpty());
   }
-  cv_.notify_all();
+
+  auto &output = ws->Output<GPUBackend>(0);
+  output.Copy(*(data.front()), (ws->has_stream() ? ws->stream() : 0));
+  // record an event so Recycle can synchronize on it
+  cudaEventRecord(cuda_event.front()->event, (ws->has_stream() ? ws->stream() : 0));
+  sync_worker_.DoWork(RecycleFunctor{ this, std::move(cuda_event), std::move(data) });
```

Comment:
Wouldn't a lambda here be enough? Or it does not work with the ownership transfers?

Reply:
It didn't work.

Reply:
Ok